File: app/src/main/java/com/dji/daw/TestComponent.java
package com.dji.daw;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
//import android.preference.PreferenceManager;
import androidx.preference.PreferenceManager;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.PopupMenu;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import dji.common.error.DJIError;
import dji.common.error.DJISDKError;
import dji.common.useraccount.UserAccountState;
import dji.common.util.CommonCallbacks;
import dji.log.DJILog;
import dji.sdk.base.BaseComponent;
import dji.sdk.base.BaseProduct;
import dji.sdk.sdkmanager.DJISDKInitEvent;
import dji.sdk.sdkmanager.DJISDKManager;
import dji.sdk.useraccount.UserAccountManager;

/**
 * Actividad que muestra opciones de prueba - para usos de programacion
 *
 */

public class TestComponent extends Activity implements View.OnClickListener, PopupMenu.OnMenuItemClickListener {
    private static final String TAG = "MainActivity";
    private static final String LAST_USED_BRIDGE_IP = "bridgeip";
    private final AtomicBoolean isRegistrationInProgress = new AtomicBoolean(false);
    private static boolean isAppStarted = false;

    private final DJISDKManager.SDKManagerCallback registrationCallback = new DJISDKManager.SDKManagerCallback() {

        @Override
        public void onRegister(DJIError error) {
            isRegistrationInProgress.set(false);
            if (error == DJISDKError.REGISTRATION_SUCCESS) {
                loginAccount();
                DJISDKManager.getInstance().startConnectionToProduct();

                Toast.makeText(getApplicationContext(), "SDK registrado", Toast.LENGTH_LONG).show();
            } else {

                Toast.makeText(getApplicationContext(),
                               "SDK no registrado",
                               Toast.LENGTH_LONG).show();
            }
        }
        @Override
        public void onProductDisconnect() {
            Toast.makeText(getApplicationContext(),
                           "Aeronave desconectada",
                           Toast.LENGTH_LONG).show();
        }

        @Override
        public void onProductConnect(BaseProduct product) {
            Toast.makeText(getApplicationContext(),
                           "Aeronave conectada",
                           Toast.LENGTH_LONG).show();
        }

        @Override
        public void onProductChanged(BaseProduct product) {

        }

        @Override
        public void onComponentChange(BaseProduct.ComponentKey key,
                                      BaseComponent oldComponent,
                                      BaseComponent newComponent) {
            Toast.makeText(getApplicationContext(),
                           key.toString() + " Detectado Cambio",
                           Toast.LENGTH_LONG).show();

        }

        @Override
        public void onInitProcess(DJISDKInitEvent event, int totalProcess) {

        }

        @Override
        public void onDatabaseDownloadProgress(long current, long total) {

        }
    };

    private void loginAccount(){
        UserAccountManager.getInstance().logIntoDJIUserAccount(this,
                new CommonCallbacks.CompletionCallbackWith<UserAccountState>() {
                    @Override
                    public void onSuccess(final UserAccountState userAccountState) {
                        Toast.makeText(getApplicationContext(),
                                "Login con DJI OK",
                                Toast.LENGTH_LONG).show();
                    }
                    @Override
                    public void onFailure(DJIError error) {
                        Toast.makeText(getApplicationContext(),
                                "Login con DJI error",
                                Toast.LENGTH_LONG).show();
                    }
                });

    }

    public static boolean isStarted() {
        return isAppStarted;
    }

    private static final String[] REQUIRED_PERMISSION_LIST = new String[] {
        Manifest.permission.VIBRATE, // Gimbal rotacion
        Manifest.permission.ACCESS_WIFI_STATE, // WIFI productos conectados
        Manifest.permission.ACCESS_COARSE_LOCATION, // Mapas
        Manifest.permission.ACCESS_NETWORK_STATE, // WIFI estado de productos conectados via WIFI
        Manifest.permission.ACCESS_FINE_LOCATION, // Mapas
        Manifest.permission.CHANGE_WIFI_STATE, // Cambio entre conexion USB o WIFI
        Manifest.permission.WRITE_EXTERNAL_STORAGE, // Log escritura
        Manifest.permission.BLUETOOTH, // Bluetooth productos conectados
        Manifest.permission.BLUETOOTH_ADMIN, // Bluetooth administracio productos conectados
        Manifest.permission.READ_EXTERNAL_STORAGE, // Log lectura
        Manifest.permission.READ_PHONE_STATE, // Identificacion del dispositivo una vez conectado
        Manifest.permission.RECORD_AUDIO // control altavoces
    };
    private static final int REQUEST_PERMISSION_CODE = 12345;
    private final List<String> missingPermission = new ArrayList<>();
    private EditText bridgeModeEditText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_testcomponent);
        isAppStarted = true;
        findViewById(R.id.complete_ui_widgets).setOnClickListener(this);
        findViewById(R.id.bt_customized_ui_widgets).setOnClickListener(this);
        findViewById(R.id.bt_map_widget).setOnClickListener(this);
        TextView versionText = (TextView) findViewById(R.id.version);
        versionText.setText(getResources().getString(R.string.sdk_version, DJISDKManager.getInstance().getSDKVersion()));
        bridgeModeEditText = (EditText) findViewById(R.id.edittext_bridge_ip);
        bridgeModeEditText.setText(PreferenceManager.getDefaultSharedPreferences(this).getString(LAST_USED_BRIDGE_IP,""));
        bridgeModeEditText.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_SEARCH
                || actionId == EditorInfo.IME_ACTION_DONE
                || event != null
                && event.getAction() == KeyEvent.ACTION_DOWN
                && event.getKeyCode() == KeyEvent.KEYCODE_ENTER) {
                if (event != null && event.isShiftPressed()) {
                    return false;
                } else {
                    // detectar ip en modo-puente si existe algun valor
                    handleBridgeIPTextChange();
                }
            }
            return false; // pasar el valor a otro servicio.
        });
        bridgeModeEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s != null && s.toString().contains("\n")) {
                    // detectar si el usaurio escribe algun texto
                    final String currentText = bridgeModeEditText.getText().toString();
                    bridgeModeEditText.setText(currentText.substring(0, currentText.indexOf('\n')));
                    handleBridgeIPTextChange();
                }
            }
        });
        checkAndRequestPermissions();
    }

    @Override
    protected void onDestroy() {
        DJISDKManager.getInstance().destroy();
        isAppStarted = false;
        super.onDestroy();
    }

    /**
     * Validar el estatus de los permisos solicitados por la  aplicacion
     */
    private void checkAndRequestPermissions() {
        // Check for permissions
        for (String eachPermission : REQUIRED_PERMISSION_LIST) {
            if (ContextCompat.checkSelfPermission(this, eachPermission) != PackageManager.PERMISSION_GRANTED) {
                missingPermission.add(eachPermission);
            }
        }
        // Request for missing permissions
        if (missingPermission.isEmpty()) {
            startSDKRegistration();
        } else {
            ActivityCompat.requestPermissions(this,
                                              missingPermission.toArray(new String[0]),
                                              REQUEST_PERMISSION_CODE);
        }
    }

    /**
     *Resultado de solicitar permisos si son permitidos
     */
    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        // Check for granted permission and remove from missing list
        if (requestCode == REQUEST_PERMISSION_CODE) {
            for (int i = grantResults.length - 1; i >= 0; i--) {
                if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                    missingPermission.remove(permissions[i]);
                }
            }
        }
        // si no hay permisos solicitarlos
        if (missingPermission.isEmpty()) {
            startSDKRegistration();
        } else {
            Toast.makeText(getApplicationContext(), "No se encuentra permisos, no se puede registrar el SDK .", Toast.LENGTH_LONG).show();
        }
    }

    private void startSDKRegistration() {
        if (isRegistrationInProgress.compareAndSet(false, true)) {
            AsyncTask.execute(() -> DJISDKManager.getInstance().registerApp(TestComponent.this, registrationCallback));
        }
    }

    @Override
    public void onClick(View view) {
        Class nextActivityClass;

        int id = view.getId();
        if (id == R.id.complete_ui_widgets) {
            nextActivityClass = WidgetPrincipalActividad.class;
        } else if (id == R.id.bt_customized_ui_widgets) {
            nextActivityClass = WidgetsActividadesPersonalizadas.class;
        } else {
            //nextActivityClass = MapWidgetActivity.class;
            PopupMenu popup = new PopupMenu(this, view);
            popup.setOnMenuItemClickListener(this);
            Menu popupMenu = popup.getMenu();
            MenuInflater inflater = popup.getMenuInflater();
            inflater.inflate(R.menu.map_select_menu, popupMenu);
            popupMenu.findItem(R.id.here_map).setEnabled(isHereMapsSupported());
            popupMenu.findItem(R.id.google_map).setEnabled(isGoogleMapsSupported(this));
            popup.show();
            return;
        }

        Intent intent = new Intent(this, nextActivityClass);
        startActivity(intent);
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public boolean onMenuItemClick(MenuItem menuItem) {
        Intent intent = new Intent(this, MapWidgetActivity.class);
        int mapBrand = 0;
        switch (menuItem.getItemId()) {
            case R.id.here_map:
                mapBrand = 0;
                break;
            case R.id.google_map:
                mapBrand = 1;
                break;
            case R.id.amap:
                mapBrand = 2;
                break;
            case R.id.mapbox:
                mapBrand = 3;
                break;
        }
        intent.putExtra(MapWidgetActivity.MAP_PROVIDER, mapBrand);
        startActivity(intent);
        return false;
    }

    public static boolean isHereMapsSupported() {
        String abi;

        abi = Build.SUPPORTED_ABIS[0];
        DJILog.d(TAG, "abi=" + abi);

        //Arquitecturas posible para procesadores strings armeabi, armeabi-v7a, arm64-v8a, x86, x86_64, mips, mips64.
        return abi.contains("arm");
    }

    public static boolean isGoogleMapsSupported(Context context) {
        GoogleApiAvailability googleApiAvailability = GoogleApiAvailability.getInstance();
        int resultCode = googleApiAvailability.isGooglePlayServicesAvailable(context);
        return resultCode == ConnectionResult.SUCCESS;
    }

    private void handleBridgeIPTextChange() {
        //El usuario ha escrito una IP y se almacena
        final String bridgeIP = bridgeModeEditText.getText().toString();

        if (!TextUtils.isEmpty(bridgeIP)) {
            DJISDKManager.getInstance().enableBridgeModeWithBridgeAppIP(bridgeIP);
            Toast.makeText(getApplicationContext(),"Modo Puente  Encendido\n IP: " + bridgeIP,Toast.LENGTH_SHORT).show();
            PreferenceManager.getDefaultSharedPreferences(this).edit().putString(LAST_USED_BRIDGE_IP,bridgeIP).apply();
        }
    }
}

File: app/src/main/java/com/dji/daw/TestComponent.kt
package com.dji.daw

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.AsyncTask
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.PopupMenu
import android.widget.TextView
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.preference.PreferenceManager
import com.dji.daw.controles.AppCompatActivityFullScreen
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import dji.common.error.DJIError
import dji.common.error.DJISDKError
import dji.common.useraccount.UserAccountState
import dji.common.util.CommonCallbacks.CompletionCallbackWith
import dji.log.DJILog
import dji.sdk.base.BaseComponent
import dji.sdk.base.BaseProduct
import dji.sdk.base.BaseProduct.ComponentKey
import dji.sdk.sdkmanager.DJISDKInitEvent
import dji.sdk.sdkmanager.DJISDKManager
import dji.sdk.sdkmanager.DJISDKManager.SDKManagerCallback
import dji.sdk.useraccount.UserAccountManager
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean

//import android.preference.PreferenceManager;
/**
 * Actividad que muestra opciones de prueba - para usos de programacion
 *
 */
class TestComponent : AppCompatActivityFullScreen(), View.OnClickListener, PopupMenu.OnMenuItemClickListener {
    private val isRegistrationInProgress = AtomicBoolean(false)
    private val registrationCallback: SDKManagerCallback = object : SDKManagerCallback {

        override fun onRegister(error: DJIError) {
            isRegistrationInProgress.set(false)
            if (error === DJISDKError.REGISTRATION_SUCCESS) {
                loginAccount()
                DJISDKManager.getInstance().startConnectionToProduct()
                Toast.makeText(applicationContext, "SDK registrado", Toast.LENGTH_LONG).show()
            } else {
                Toast.makeText(applicationContext,
                        "SDK no registrado",
                        Toast.LENGTH_LONG).show()
            }
        }

        override fun onProductDisconnect() {
            Toast.makeText(applicationContext,
                    "Aeronave desconectada",
                    Toast.LENGTH_LONG).show()
        }

        override fun onProductConnect(product: BaseProduct) {
            Toast.makeText(applicationContext,
                    "Aeronave conectada",
                    Toast.LENGTH_LONG).show()
        }

        override fun onProductChanged(product: BaseProduct) {}
        override fun onComponentChange(key: ComponentKey?,
                                       oldComponent: BaseComponent?,
                                       newComponent: BaseComponent?) {
            Toast.makeText(applicationContext,
                    "$key Detectado Cambio",
                    Toast.LENGTH_LONG).show()
        }

        override fun onInitProcess(event: DJISDKInitEvent, totalProcess: Int) {}
        override fun onDatabaseDownloadProgress(current: Long, total: Long) {}
    }

    private fun loginAccount() {
        UserAccountManager.getInstance().logIntoDJIUserAccount(this,
                object : CompletionCallbackWith<UserAccountState?> {
                    override fun onSuccess(userAccountState: UserAccountState?) {
                        Toast.makeText(applicationContext,
                                "Login con DJI OK",
                                Toast.LENGTH_LONG).show()
                    }

                    override fun onFailure(error: DJIError) {
                        Toast.makeText(applicationContext,
                                "Login con DJI error",
                                Toast.LENGTH_LONG).show()
                    }
                })
    }

    private val missingPermission: MutableList<String> = ArrayList()
    private var bridgeModeEditText: EditText? = null


    /// aqui inicia
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_testcomponent)
        isStarted = true
        findViewById<View>(R.id.complete_ui_widgets).setOnClickListener(this)
        findViewById<View>(R.id.bt_customized_ui_widgets).setOnClickListener(this)
        findViewById<View>(R.id.bt_map_widget).setOnClickListener(this)
        val versionText = findViewById<View>(R.id.version) as TextView
        versionText.text = resources.getString(R.string.sdk_version, DJISDKManager.getInstance().sdkVersion)
        bridgeModeEditText = findViewById<View>(R.id.edittext_bridge_ip) as EditText
        bridgeModeEditText!!.setText(PreferenceManager.getDefaultSharedPreferences(this).getString(LAST_USED_BRIDGE_IP, ""))
        bridgeModeEditText!!.setOnEditorActionListener { v: TextView?, actionId: Int, event: KeyEvent? ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH || actionId == EditorInfo.IME_ACTION_DONE || event != null && event.action == KeyEvent.ACTION_DOWN && event.keyCode == KeyEvent.KEYCODE_ENTER) {
                if (event != null && event.isShiftPressed) {
                    return@setOnEditorActionListener false
                } else {
                    // detectar ip en modo-puente si existe algun valor
                    handleBridgeIPTextChange()
                }
            }
            false // pasar el valor a otro servicio.
        }
        bridgeModeEditText!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                if (s.toString().contains("\n")) {
                    // detectar si el usaurio escribe algun texto
                    val currentText = bridgeModeEditText!!.text.toString()
                    bridgeModeEditText!!.setText(currentText.substring(0, currentText.indexOf('\n')))
                    handleBridgeIPTextChange()
                }
            }
        })
        checkAndRequestPermissions()
    }

    override fun onDestroy() {
        DJISDKManager.getInstance().destroy()
        isStarted = false
        super.onDestroy()
    }

    /**
     * Validar el estatus de los permisos solicitados por la  aplicacion
     */
    private fun checkAndRequestPermissions() {
        // Check for permissions
        for (eachPermission in REQUIRED_PERMISSION_LIST) {
            if (ContextCompat.checkSelfPermission(this, eachPermission) != PackageManager.PERMISSION_GRANTED) {
                missingPermission.add(eachPermission)
            }
        }
        // Request for missing permissions
        if (missingPermission.isEmpty()) {
            startSDKRegistration()
        } else {
            ActivityCompat.requestPermissions(this,
                    missingPermission.toTypedArray(),
                    REQUEST_PERMISSION_CODE)
        }
    }

    /**
     * Resultado de solicitar permisos si son permitidos
     */
    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>,
                                            grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        // Check for granted permission and remove from missing list
        if (requestCode == REQUEST_PERMISSION_CODE) {
            for (i in grantResults.indices.reversed()) {
                if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                    missingPermission.remove(permissions[i])
                }
            }
        }
        // si no hay permisos solicitarlos
        if (missingPermission.isEmpty()) {
            startSDKRegistration()
        } else {
            Toast.makeText(applicationContext, "No se encuentra permisos, no se puede registrar el SDK .", Toast.LENGTH_LONG).show()
        }
    }

    private fun startSDKRegistration() {
        if (isRegistrationInProgress.compareAndSet(false, true)) {
            AsyncTask.execute { DJISDKManager.getInstance().registerApp(this@TestComponent, registrationCallback) }
        }
    }

    override fun onClick(view: View) {
        val nextActivityClass: Class<*>
        val id = view.id
        if (id == R.id.complete_ui_widgets) {
            nextActivityClass = WidgetPrincipalActividad::class.java
        } else if (id == R.id.bt_customized_ui_widgets) {
            nextActivityClass = WidgetsActividadesPersonalizadas::class.java
        } else {
            nextActivityClass = MapWidgetActivity::class.java
            val popup = PopupMenu(this, view)
            popup.setOnMenuItemClickListener(this)
            val popupMenu = popup.menu
            val inflater = popup.menuInflater
            inflater.inflate(R.menu.map_select_menu, popupMenu)
            popupMenu.findItem(R.id.here_map).isEnabled = isHereMapsSupported
            popupMenu.findItem(R.id.google_map).isEnabled = isGoogleMapsSupported(this)
            popup.show()
            return
        }
        val intent = Intent(this, nextActivityClass)
        startActivity(intent)
    }

    @SuppressLint("NonConstantResourceId")
    override fun onMenuItemClick(menuItem: MenuItem): Boolean {
        val intent = Intent(this, MapWidgetActivity::class.java)
        var mapBrand = 0
        when (menuItem.itemId) {
            R.id.here_map -> mapBrand = 0
            R.id.google_map -> mapBrand = 1
            R.id.amap -> mapBrand = 2
            R.id.mapbox -> mapBrand = 3
        }
        intent.putExtra(MapWidgetActivity.MAP_PROVIDER, mapBrand)
        startActivity(intent)
        return false
    }

    private fun handleBridgeIPTextChange() {
        //El usuario ha escrito una IP y se almacena
        val bridgeIP = bridgeModeEditText!!.text.toString()
        if (!TextUtils.isEmpty(bridgeIP)) {
            DJISDKManager.getInstance().enableBridgeModeWithBridgeAppIP(bridgeIP)
            Toast.makeText(applicationContext, "Modo Puente  Encendido\n IP: $bridgeIP", Toast.LENGTH_SHORT).show()
            PreferenceManager.getDefaultSharedPreferences(this).edit().putString(LAST_USED_BRIDGE_IP, bridgeIP).apply()
        }
    }


    companion object {
        private const val TAG = "MainActivity"
        private const val LAST_USED_BRIDGE_IP = "bridgeip"
        var isStarted = false
            private set
        private val REQUIRED_PERMISSION_LIST = arrayOf(
                Manifest.permission.VIBRATE,  // Gimbal rotacion
                Manifest.permission.ACCESS_WIFI_STATE,  // WIFI productos conectados
                Manifest.permission.ACCESS_COARSE_LOCATION,  // Mapas
                Manifest.permission.ACCESS_NETWORK_STATE,  // WIFI estado de productos conectados via WIFI
                Manifest.permission.ACCESS_FINE_LOCATION,  // Mapas
                Manifest.permission.CHANGE_WIFI_STATE,  // Cambio entre conexion USB o WIFI
                Manifest.permission.WRITE_EXTERNAL_STORAGE,  // Log escritura
                Manifest.permission.BLUETOOTH,  // Bluetooth productos conectados
                Manifest.permission.BLUETOOTH_ADMIN,  // Bluetooth administracio productos conectados
                Manifest.permission.READ_EXTERNAL_STORAGE,  // Log lectura
                Manifest.permission.READ_PHONE_STATE,  // Identificacion del dispositivo una vez conectado
                Manifest.permission.RECORD_AUDIO // control altavoces
        )
        private const val REQUEST_PERMISSION_CODE = 12345

        //Arquitecturas posible para procesadores strings armeabi, armeabi-v7a, arm64-v8a, x86, x86_64, mips, mips64.
        val isHereMapsSupported: Boolean
            get() {
                val abi: String
                abi = Build.SUPPORTED_ABIS[0]
                DJILog.d(TAG, "abi=$abi")

                //Arquitecturas posible para procesadores strings armeabi, armeabi-v7a, arm64-v8a, x86, x86_64, mips, mips64.
                return abi.contains("arm")
            }

        fun isGoogleMapsSupported(context: Context?): Boolean {
            val googleApiAvailability = GoogleApiAvailability.getInstance()
            val resultCode = googleApiAvailability.isGooglePlayServicesAvailable(context)
            return resultCode == ConnectionResult.SUCCESS
        }
    }
}